PyPy still crash! (only with pytest & co)

CI failure: https://github.com/diofant/python-gmp/actions/runs/12270822891/job/34236679337

```
============================= test session starts ==============================
platform linux -- Python 3.10.14[pypy-7.3.18-alpha], pytest-8.3.4, pluggy-1.5.0 -- /home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/bin/python
cachedir: .pytest_cache
hypothesis profile 'ci' -> database=None, deadline=None, print_blob=True, derandomize=True, max_examples=10000, suppress_health_check=[HealthCheck.too_slow]
rootdir: /home/runner/work/python-gmp/python-gmp
configfile: pyproject.toml
plugins: hypothesis-6.122.3
collecting ... collected 34 items

tests/test_functions.py::test_isqrt PASSED                               [  2%]
tests/test_functions.py::test_gcd PASSED                                 [  5%]
tests/test_mpz.py::test_mpz_from_to_str PASSED                           [  8%]
tests/test_mpz.py::test_mpz_from_to_int PASSED                           [ 11%]
tests/test_mpz.py::test_repr PASSED                                      [ 14%]
tests/test_mpz.py::test_richcompare PASSED                               [ 17%]
tests/test_mpz.py::test_mpz_hash PASSED                                  [ 20%]
tests/test_mpz.py::test_mpz_plus_minus_abs PASSED                        [ 23%]
tests/test_mpz.py::test_add_sub PASSED                                   [ 26%]
tests/test_mpz.py::test_mul PASSED                                       [ 29%]
tests/test_mpz.py::test_divmod PASSED                                    [ 32%]
tests/test_mpz.py::test_power PASSED                                     [ 35%]
tests/test_mpz.py::test_invert PASSED                                    [ 38%]
tests/test_mpz.py::test_and PASSED                                       [ 41%]
tests/test_mpz.py::test_or PASSED                                        [ 44%]
tests/test_mpz.py::test_xor PASSED                                       [ 47%]
tests/test_mpz.py::test_lshift PASSED                                    [ 50%]
tests/test_mpz.py::test_rshift PASSED                                    [ 52%]
tests/test_mpz.py::test_getseters PASSED                                 [ 55%]
tests/test_mpz.py::test_methods PASSED                                   [ 58%]
tests/test_mpz.py::test_to_bytes XFAIL (https://github.com/diofant/p...) [ 61%]
tests/test_mpz.py::test_from_bytes XFAIL (https://github.com/diofant...) [ 64%]
tests/test_mpz.py::test___float__ XFAIL (https://github.com/diofant/...) [ 67%]
tests/test_mpz.py::test___round__ XFAIL (https://github.com/diofant/...) [ 70%]
tests/test_mpz.py::test___sizeof__ SKIPPED (sys.getsizeof raises Typ...) [ 73%]
tests/test_mpz.py::test_digits SKIPPED (could not import 'gmpy2': No...) [ 76%]
tests/test_mpz.py::test_digits_frombase_low PASSED                       [ 79%]
tests/test_mpz.py::test_digits_frombase_high SKIPPED (could not impo...) [ 82%]
tests/test_mpz.py::test_frombase_auto PASSED                             [ 85%]
tests/test_mpz.py::test_pickle[2] PASSED                                 [ 88%]
tests/test_mpz.py::test_pickle[3] PASSED                                 [ 91%]
tests/test_mpz.py::test_pickle[4] PASSED                                 [ 94%]
tests/test_mpz.py::test_pickle[5] PASSED                                 [ 97%]
Fatal Python error: Aborted

Stack (most recent call first, approximate line numbers):
  File "/home/runner/work/python-gmp/python-gmp/tests/test_mpz.py", line 412 in test_outofmemory
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/python.py", line 152 in pytest_pyfunc_call
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_callers.py", line 53 in _multicall
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_manager.py", line 111 in _hookexec
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_hooks.py", line 498 in __call__
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/python.py", line 1625 in runtest
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/runner.py", line 163 in pytest_runtest_call
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_callers.py", line 53 in _multicall
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_manager.py", line 111 in _hookexec
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_hooks.py", line 498 in __call__
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/runner.py", line 242 in <lambda>
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/runner.py", line 319 in from_call
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/runner.py", line 226 in call_and_report
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/runner.py", line 118 in runtestprotocol
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/runner.py", line 110 in pytest_runtest_protocol
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_callers.py", line 53 in _multicall
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_manager.py", line 111 in _hookexec
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_hooks.py", line 498 in __call__
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/main.py", line 350 in pytest_runtestloop
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_callers.py", line 53 in _multicall
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_manager.py", line 111 in _hookexec
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_hooks.py", line 498 in __call__
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/main.py", line 333 in _main
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/main.py", line 270 in wrap_session
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/main.py", line 329 in pytest_cmdline_main
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_callers.py", line 53 in _multicall
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_manager.py", line 111 in _hookexec
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/pluggy/_hooks.py", line 498 in __call__
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/config/__init__.py", line 139 in main
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/lib/pypy3.10/site-packages/_pytest/config/__init__.py", line 194 in console_main
  File "/home/runner/work/_temp/e3037a55-e39d-44dc-beaa-60dbd1a32f03/pypy-c-jit-185785-d81836603e77-linux64/bin/pytest", line 1 in <module>
  File "<builtin>/app_main.py", line 1053 in execfile
  File "<builtin>/app_main.py", line 140 in run_toplevel
  File "<builtin>/app_main.py", line 759 in run_command_line
  File "<builtin>/app_main.py", line 1162 in entry_point
/home/runner/work/_temp/3fff063e-8b7e-4a05-9b5e-62ae109d6272.sh: line 1:  2454 Aborted                 (core dumped) pytest
tests/test_mpz.py::test_outofmemory 
Error: Process completed with exit code 134.
```

--- a/tests/test_memory.py
+++ b/tests/test_memory.py
@@ -1,9 +1,9 @@
+import gc
 import platform
+import random
 
 import pytest
 from gmp import fac, mpz
-from hypothesis import given
-from hypothesis.strategies import integers
 
 if platform.system() != "Linux":
     pytest.skip("FIXME: setrlimit fails with ValueError on MacOS",
@@ -11,37 +11,45 @@ if platform.system() != "Linux":
 if platform.python_implementation() == "GraalVM":
     pytest.skip("XXX: module 'resource' has no attribute 'setrlimit'",
                 allow_module_level=True)
-if platform.python_implementation() == "PyPy":
-    pytest.skip("XXX: diofant/python-gmp#73", allow_module_level=True)
+if platform.python_implementation() != "PyPy":
+    VMEM_LIMIT = 64*1000**2
+else:
+    VMEM_LIMIT = 128*1000**2
 
 resource = pytest.importorskip("resource")
 
 
-@given(integers(min_value=12811, max_value=24984))
-def test_fac_outofmem(x):
-    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
-    resource.setrlimit(resource.RLIMIT_AS, (1024*32*1024, hard))
-    a = mpz(x)
-    while True:
-        try:
-            fac(a)
-            a *= 2
-        except MemoryError:
-            break
-    resource.setrlimit(resource.RLIMIT_AS, (soft, hard))
+def test_fac_outofmem():
+    for _ in range(100):
+        x = random.randint(12811, 24984)
+        soft, hard = resource.getrlimit(resource.RLIMIT_AS)
+        resource.setrlimit(resource.RLIMIT_AS, (VMEM_LIMIT, hard))
+        a = mpz(x)
+        while True:
+            try:
+                fac(a)
+                a *= 2
+            except MemoryError:
+                del a
+                gc.collect()
+                break
+        resource.setrlimit(resource.RLIMIT_AS, (soft, hard))
 
 
-@given(integers(min_value=49846727467293, max_value=249846727467293))
-def test_square_outofmem(x):
-    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
-    resource.setrlimit(resource.RLIMIT_AS, (1024*32*1024, hard))
-    mx = mpz(x)
-    i = 1
-    while True:
-        try:
-            mx = mx*mx
-        except MemoryError:
-            assert i > 5
-            break
-        i += 1
-    resource.setrlimit(resource.RLIMIT_AS, (soft, hard))
+def test_square_outofmem():
+    for _ in range(100):
+        x = random.randint(49846727467293, 249846727467293)
+        soft, hard = resource.getrlimit(resource.RLIMIT_AS)
+        resource.setrlimit(resource.RLIMIT_AS, (VMEM_LIMIT, hard))
+        mx = mpz(x)
+        i = 1
+        while True:
+            try:
+                mx = mx*mx
+            except MemoryError:
+                del mx
+                gc.collect()
+                assert i > 5
+                break
+            i += 1
+        resource.setrlimit(resource.RLIMIT_AS, (soft, hard))

